### DISCUSSION TITLE
Unit multi-language simple docker container

### DISCUSSION THREAD

**Discussion by javorszky** [nginx/unit#1264] [Ideas]
Hello!

We've been trying to put together a docker image that supports multiple images in a way that's also easy to understand and easy to replicate. Here's where the current state of this is:

## Existing images

We provide a few docker images for single language supports or no language support in our repository in the [pkg/docker](https://github.com/nginx/unit/tree/master/pkg/docker) directory.

Each of the language specific image is based on a docker image for that specific language. In PHP 8.3's case, it starts with this:

```dockerfile
FROM php:8.3-cli-bullseye
```

The only outlier is the `-minimal` unit image which starts with `debian:bullseye-slim`.

## Aim

We're looking to create an image that supports PHP, Python, and WebAssembly out of the box in a single image.

That single image should be easy to develop and understand its inner workings.

Ideally it should either be based on an image where unit is already installed, or unit should be installed easily with something like `apt install unit` or `apk add unit`.

The point is that instead of having incredibly long sections where we configure and build and configure and build unit and related modules from source, we have individual lines of adding the required software from a package manager of the underlying system and we should be ready to go.

## Non-goals

Final docker image size is not something I want to care about at the moment.

### Based on `debian:bookworm-slim`

`unit` is not available to install as a package. This means we'd have to copy in the entire source code and build it from scratch, which is not something we want to do.

### Based on `unit:minimal`

It's based on `bullseye-slim` (debian 11), where the latest PHP version is 7.4. The slim image out of the box doesn't work when following this guide: https://php.watch/articles/php-8.3-install-upgrade-on-debian-ubuntu#detailed

The `lsb_release` command is not available in the slim image, which results in a malformed entry in the sources file. Fixing that with `apt install lsb-release` enables installation of php8.3.

Python's version 3.9 is installed, however to install 3.11 or 3.12 we'd need to install python from source, which is not ideal.

### Based on the `alpine` image

Wasmtime would need to be built from source, but the C compiler version is the other one that is in wasmtime's toolchain.

Otherwise PHP and Python are both available, as well as unit and the language specific extensions on it.

## Recap

What do we need to do on our end to make sure we can build a multi-language docker image easily such that a user who knows some docker, some linux package management, but very little about compiling software from source can replicate and customize what we're doing in the image. A pseudo-dockerfile is below:

```
FROM debian

RUN apt update -y && apt upgrade -y

RUN apt install -y php8.3 python3 wasmtime

RUN apt install unit unit-php unit-python unit-wasm

CMD ['unitd', '--no-daemon', '--control', 'unix:/var/run/control.unit.sock']
```

**Comment by callahad** [comment]
I'd maybe question this aim:

> Ideally it should either be based on an image where unit is already installed, or unit should be installed easily with something like `apt install unit` or `apk add unit`.

We're going to need an easy path to build images with development sources of Unit. Is there a way we can use this effort to _also_ achieve that?

**Reply by javorszky** [reply]
We can create a docker image that would take a directory in a known place and compile that version of unit into the image, and then use that image as a base in others.

```
/directory
  - dockerfile
  - unit-source/
  - makefile
```
The dockerfile would look like this:
```dockerfile
FROM bullseye-slim

...

WORKDIR /src/unit/
COPY ./unit-source/ .

<compile unit in the directory>
```
Then you can tag that as `unit:dev` and in the pseudo-dockerfile above, you can swap the `apt install unit` part with `FROM unit:dev`, like so:
```dockerfile
FROM unit:dev

RUN apt update -y && apt upgrade -y

RUN apt install -y php8.3 python3 wasmtime

RUN apt install unit-php unit-python unit-wasm // these probably would need to be built from source as well?

CMD ['unitd', '--no-daemon', '--control', 'unix:/var/run/control.unit.sock']
```

**Reply by callahad** [reply]
Yeah. Or a multi-stage build that does a `COPY --from`, even.

**Reply by javorszky** [reply]
I also have an idea about a UI where we can use a bunch of questions in a flowchart that would result in a set of dockerfiles and instructions based on what people want to achieve.

**Reply by javorszky** [reply]
> Or a multi-stage build that does a COPY --from, even.

Yes, though I haven't figured out how to do a multi-stage build where a `FROM ... AS step 4` is a dockerfile rather than a pre-built image someplace 🤔

**Reply by lcrilly** [reply]
I believe multi-stage builds are not permitted for Docker Official Images. @thresheek is the expert on this topic.

**Reply by javorszky** [reply]
> I believe multi-stage builds are not permitted for Docker Official Images

That's another question we talked about during our standup on 14th may: do we really _need_ this image to be on the official images repo, or are we okay with this being under nginx/unit? The benefits on the surface outweigh the nicety of having to use nginx/ in the image name.

**Reply by lcrilly** [reply]
Let's understand the constraints first. To enjoy the benefits of the new image, you first need to discover and acquire the image. Nothing beats a Docker Official Image for discoverability and confidence.

**Reply by callahad** [reply]
@lcrilly We know we need a Ridiculously Easy omnibus image as a first port of call. If we can achieve that as an Official Image, great! If not, we still need the omnibus image. So we'll start by building that, and then figure out if we can back it into the Official Image programme.

**Comment by callahad** [comment]
I'm a little leery of elevating downstream distro packages of Unit (or even of Python / PHP / Wasmtime) into being the foundation of our upstream Docker images.

What do we do when Python n+1 releases, but isn't available in Debian / Alpine?

**Reply by javorszky** [reply]
That's the exact issue we have with php8.3 and debian. It's a value judgement and decision we as the unit team need to make:
* do we make our jobs easier and say "here's a bunch of example docker images, essentially you need to learn how to build unit and python from source and look up the articles that actually work in adding python version n+1 to debian/alpine, good luck!"
* or do we do this work to make our users' lives easier? Even if it's not a docker image, but maybe some documentation that we keep reviewing to make sure that what we've written there still works

**Comment by callahad** [comment]
Do we know how we want to approach to runtime versions for the omnibus image? Latest stable for everything? Multiple language versions all in the same image? Combinatoric explosion of images?

E.g., I don't think it'd be unreasonable to want (Py311, Py312) × (PHP82, PHP83), etc.

**Reply by javorszky** [reply]
I think one image of the latest stable versions of the languages is a good opinionated approach. Normally if someone can install php8.3, then php8.2 is also available. Same with python3.12, if that's available, so is 3.11, and usually in that case it's a case of swapping version numbers during installation of those languages.

**Reply by callahad** [reply]
Aside: If Python + PHP proves painful, a good escape hatch would be two images: Python or PHP. Both with Wasmtime. I think the key direction we need to move in is making Wasm support an integral feature of all Unit releases.

**Reply by avahahn** [reply]
could we could just statically link the wasm components at build time?
Then that specific goal would only a matter of extending the build processes of the existing unit docker images.

**Reply by ac000** [reply]
If you go with Debian and install the unit-wasm package then it comes with a wasmtime DSO.

```
# dpkg -L unit-wasm
/.
/usr
/usr/lib
/usr/lib/libwasmtime.so
/usr/lib/unit
/usr/lib/unit/debug-modules
/usr/lib/unit/debug-modules/wasm.unit.so
/usr/lib/unit/debug-modules/wasm_wasi_component.unit.so
/usr/lib/unit/modules
/usr/lib/unit/modules/wasm.unit.so
/usr/lib/unit/modules/wasm_wasi_component.unit.so
/usr/share
/usr/share/doc
/usr/share/doc/unit-wasm
/usr/share/doc/unit-wasm/changelog.Debian.gz
/usr/share/doc/unit-wasm/changelog.gz
/usr/share/doc/unit-wasm/copyright
```

But if you only need `wasm_wasi_component.unit.so`, that is the component supporting rust based language module. Then you don't even need the wasmtime DSO.

**Comment by ac000** [comment]
> Based on debian:bookworm-slim

> unit is not available to install as a package. This means we'd have to copy in the entire source code and build it from scratch, which is not something we want to do.

Hmm, unless I'm missing something, _we_ provide Debian Bookworm packages, take a browse through [here](https://packages.nginx.org/unit/debian/pool/unit/u/)

**Reply by javorszky** [reply]
Based on that, my expectation is that when I'm inside this image:
```dockerfile
FROM debian:bookworm-slim

RUN apt update && apt upgrade
```
I could run `apt install unit`, and I would have unit installed. This happens however:

```
# apt install unit
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package unit
#
```
What's the correct package name to install via apt inside a debian container?

**Reply by callahad** [reply]
From @ac000:

> _We_ provide the packages, as we do for many distributions. In this case you need to install the apt source, see https://unit.nginx.org/installation/#debian

(Reposting here because emailed replies don't thread properly in GitHub Discussions)

**Comment by abbas542** [comment]
> Hello!

We've been trying to put together a docker image that supports multiple images in a way that's also easy to understand and easy to replicate. Here's where the current state of this is:

Existing images
We provide a few docker images for single language supports or no language support in our repository in the [pkg/docker](https://github.com/nginx/unit/tree/master/pkg/docker) directory.

Each of the language specific image is based on a docker image for that specific language. In PHP 8.3's case, it starts with this:

FROM php:8.3-cli-bullseye
The only outlier is the -minimal unit image which starts with debian:bullseye-slim.

Aim
We're looking to create an image that supports PHP, Python, and WebAssembly out of the box in a single image.

That single image should be easy to develop and understand its inner workings.

Ideally it should either be based on an image where unit is already installed, or unit should be installed easily with something like apt install unit or apk add unit.

The point is that instead of having incredibly long sections where we configure and build and configure and build unit and related modules from source, we have individual lines of adding the required software from a package manager of the underlying system and we should be ready to go.

Non-goals
Final docker image size is not something I want to care about at the moment.

Issues we've faced so far
Based on debian:bookworm-slim
unit is not available to install as a package. This means we'd have to copy in the entire source code and build it from scratch, which is not something we want to do.

Based on unit:minimal
It's based on bullseye-slim (debian 11), where the latest PHP version is 7.4. The slim image out of the box doesn't work when following this guide: https://php.watch/articles/php-8.3-install-upgrade-on-debian-ubuntu#detailed

The lsb_release command is not available in the slim image, which results in a malformed entry in the sources file. Fixing that with apt install lsb-release enables installation of php8.3.

Python's version 3.9 is installed, however to install 3.11 or 3.12 we'd need to install python from source, which is not ideal.

Based on the alpine image
Wasmtime would need to be built from source, but the C compiler version is the other one that is in wasmtime's toolchain.

Otherwise PHP and Python are both available, as well as unit and the language specific extensions on it.

Recap
What do we need to do on our end to make sure we can build a multi-language docker image easily such that a user who knows some docker, some linux package management, but very little about compiling software from source can replicate and customize what we're doing in the image. A pseudo-dockerfile is below:

FROM debian

RUN apt update -y && apt upgrade -y

RUN apt install -y php8.3 python3 wasmtime

RUN apt install unit unit-php unit-python unit-wasm

CMD ['unitd', '--no-daemon', '--control', 

 'unix:/var/run/control.unit.sock'] 




Your goal of creating a simple, multi-language Docker image is great for usability! Key suggestions:

Base Image Choice: Consider a base like debian:bookworm or ubuntu:22.04, which support newer versions of PHP and Python through PPAs or backports[.](https://menuland.ph/wendys-menu-in-the-philippines-2024/)
Install Scripts: Use well-documented package installation commands to keep it user-friendly.
Unit Packages: Explore custom repositories or community-maintained packages to avoid compiling unit from source.
